[docs] [refactor] Add help messages for datadir path mangling

jamesob opened the pull request:
Change `-conf`'s help message to indicate that relative path values will be prefixed by the datadir path. This behavior probably merits clarification; it's kind of confusing when attempting to specify a configuration file in the current directory with `-conf=bitcoin.conf`, but instead loading the `bitcoin.conf` file in ~/.bitcoin datadir.

### Edit

This PR has been modified to document all cases where relative path configurations are modified to be under datadir. A small refactoring has also been added which consolidates this normalization.

laanwj commented:
Not just this one, all relative files and directories specified either on the command line or through RPC are supposed to be relative to the data directory.

MarcoFalke commented:
Wasn't walletdir temporarily allowed to be relative to the cwd?

jamesob commented:
Sounds like we don't have any documentation indicating that all path arguments are relative to datadir;  that information should live somewhere (preferably in a single place that isn't easy to miss). One option is to put it in a paragraph between "Usage" and "Options" in `{bitcoind,bitcoin-cli,bitcoin-qt} -h`. Any other suggestions?

MarcoFalke commented:
Options that take paths or file names are:

* `-rpccookiefile=<loc>`
* `-walletdir=<dir>`
* `-wallet=<file>`
* `-datadir=<dir>`
* `-conf=<file>`

So, adding a general paragraph only makes sense if it holds true for all paths. Otherwise, I'd prefer if each command line help was clarified individually.

jamesob committed:
Add AbsPathForConfigVal to consolidate datadir prefixing for path args

Most commandline/config args are interpreted as relative to datadir if
not passed absolute. Consolidate the logic for this normalization.

jamesob commented:
@MarcoFalke thanks for that list. I've looked at how each option on the list is parsed (as well as a few others):

- `-rpccookiefile=<loc>`: Is put under datadir if relative.
- `-walletdir=<dir>`: Isn't allowed to be relative.
- `-wallet=<file>`: Moved under walletdir with `fs::absolute(walletFile, GetWalletDir())`.
- `-datadir=<dir>`: Can be specified as relative but will warn. Obviously not put under itself.
- `-conf=<file>`: Is put under datadir if relative.
- `-loadblock=<file>`: Interpreted literally -- isn't modified if relative. Passed directly to `fsbridge::fopen`.
- `-debuglogfile=<file>`: Is put under datadir if relative.
- `-pid=<file>`: Is put under datadir if relative. 

So: each of these options is either not allowed to be relative or is put under datadir if relative with the exception of `wallet` and `loadblock`. Because we don't uniformly treat relative paths as relative to datadir, I've updated only the affected help messages.

I've also added a refactoring commit which consolidates the "normalization" of relative paths under datadir.

laanwj commented:
Thanks, utACK

promag commented on src/util.cpp:
```diff
@@ -936,3 +927,10 @@ int64_t GetStartupTime()
     return nStartupTime;
 }
+
+fs::path NormalizePathArg(const fs::path& pathArg, bool fNetSpecific)
```
I'd rename to `GetAbsolutePath` or more specific `GetAbsoluteDataPath`:
```cpp
fs::path GetAbsoluteDataPath(const fs::path& path, bool net_specific)
```
I other contexts/languages, normalization can be interpreted as transforming from `foo/../bar/` to `bar`.

Nit, snake_case arguments.

promag commented on src/wallet/init.cpp:
```diff
@@ -7,6 +7,7 @@
 
 #include <net.h>
 #include <util.h>
+#include <fs.h>
```
Nit, unrelated change, remove.

promag reviewed (commented):
Concept ACK.

jamesob commented:
Thanks for the feedback @promag. It's been incorporated and pushed.

ryanofsky commented on src/util.cpp:
```diff
@@ -936,3 +927,10 @@ int64_t GetStartupTime()
+fs::path GetAbsolutePathForArg(const fs::path& path, bool net_specific)
+{
+    if (!path.is_complete())
+        return GetDataDir(net_specific) / path;
```
Would be a little better to return: `return fs::absolute(path, GetDataDir(net_specific))` to make it clear this returns an absolute path. You could also drop the `is_complete` check above if you did this.

jamesob replied:
Great, thanks.

ryanofsky commented on src/util.cpp:
```diff
@@ -936,3 +927,10 @@ int64_t GetStartupTime()
+
+fs::path GetAbsolutePathForArg(const fs::path& path, bool net_specific)
+{
+    if (!path.is_complete())
```
Could eliminate this check, but if you will keep it [developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md) suggest adding braces or putting if body on same line as condition.

jamesob replied:
Taking your advice below, but the reminder about convention is appreciated.

ryanofsky commented on src/util.h:
```diff
@@ -191,6 +191,16 @@ bool OpenDebugLog();
+ * @param net_specific Forwarded to GetDataDir().
+ * @return The normalized path.
+ */
+fs::path GetAbsolutePathForArg(const fs::path& path, bool net_specific = true);
```
"ForArg" in function name seems a little redundant. Just a suggestion, but I might call it something `GetDataPath` or `GetDataDirPath`.

jamesob replied:
My only reservation here is that the path won't necessarily be in datadir, so a name with Data or DataDir seems confusing. "Arg" here is meant to correspond to an argument to the binary, not the function, but maybe that isn't clear.

jamesob replied:
I'll try `AbsPathForConfigVal`.

ryanofsky reviewed (commented):
utACK 23d295aa9e16cb3591bbd8bb0f5ce026b861b7b6

jamesob commented:
Thanks for the good feedback, @ryanofsky. Incorporated and pushed.

ajtowns commented:
Could update the commit title to match the renamed function :)

jamesob commented:
d'oh. Thanks, @ajtowns. Pushed.

jnewbery commented:
Looks good.

Minor note that `-conf` is relative to the base datadir, but `-debuglogfile`, `-pid` and `-rpccookiefile` are relative the net-specific subdirectories. It might be better to specify that in the help text.

sipa commented:
utACK 95963c120c48ce750fdb2298f3eb70034e578e8d, but needs rebase.

jamesob committed:
Clarify help messages for path args to mention datadir prefix

Change `-conf`'s and others' help messages to indicate that relative path
values will be prefixed by the datadir path. This behavior is confusing when
attempting to specify a configuration file in the current directory with
`-conf=bitcoin.conf`, but loading the `bitcoin.conf` file in ~/.bitcoin
datadir.

jamesob commented:
Rebased and incorporated @jnewbery's feedback.

promag commented on src/util.h:
```diff
@@ -191,6 +191,16 @@ bool OpenDebugLog();
+ *
+ * @param path The path to be conditionally prefixed with datadir.
+ * @param net_specific Forwarded to GetDataDir().
+ * @return The normalized path.
```
The absolute path?

jnewbery commented:
utACK 54604600c3de6cb18540c0911127173f68ad246c

MarcoFalke commented:
Post-merge utACK 54604600c3de6cb18540c0911127173f68ad246c
